core/tests/manual/main.rs

```diff
@@ -53,14 +67,98 @@ impl CodeBlockType {
     }
 }
 
-fn check_repl_parts(content: String) {
+#[derive(Debug, Clone)]
+enum MessageExpectation {
+    Full(String),
+    Abridged(String),
+}
+
+#[derive(Debug, Clone)]
+enum ReplResult {
+    Value(String),
+    Error(MessageExpectation),
+    Empty,
+}
+
+fn extract_repl_piece(piece: impl AsRef<str>) -> (String, ReplResult) {
+    let lines: Vec<&str> = piece.as_ref().split_inclusive('\n').collect();
+    let split_index = lines
+        .iter()
+        .position(|l| !l.starts_with(' '))
+        .unwrap_or(lines.len());
+    let (program_lines, result_lines) = lines.split_at(split_index);
+
+    let result_string = result_lines.concat();
+    let result = if result_string.is_empty() {
+        ReplResult::Empty
+    } else if result_string.starts_with("error:") {
+        if let Some((result_string, _)) = result_string.rsplit_once("[...]") {
+            ReplResult::Error(MessageExpectation::Abridged(result_string.to_owned()))
+        } else {
+            ReplResult::Error(MessageExpectation::Full(result_string))
+        }
+    } else {
+        ReplResult::Value(result_string)
+    };
+
+    (program_lines.concat(), result)
+}
+
+#[track_caller]
+fn assert_str_eq_approx(actual: impl AsRef<str>, expected: impl AsRef<str>) {
+    assert_str_eq!(
+        actual
+            .as_ref()
+            .lines()
+            .flat_map(|l| once(l.trim_end()).chain(once("\n")))
+            .collect::<String>()
+            .trim_end(),
+        expected.as_ref().trim_end()
+    );
+}
+
+#[track_caller]
+fn assert_prefix(actual: impl AsRef<str>, prefix: impl AsRef<str>) {
+    assert!(
+        actual.as_ref().starts_with(prefix.as_ref()),
+        "{} was expected to be a prefix of {}",
+        prefix.as_ref(),
+        actual.as_ref()
+    );
+}
+
+fn check_error_report(actual: impl AsRef<str>, expected: MessageExpectation) {
+    match expected {
+        MessageExpectation::Full(expected) => assert_str_eq_approx(actual, expected),
+        MessageExpectation::Abridged(prefix) => assert_prefix(actual, prefix),
+    }
+}
+
+fn check_repl(content: String) {
+    let mut repl = ReplImpl::<CacheImpl>::new(std::io::sink());
+    repl.load_stdlib().unwrap();
     for piece in content.split("\n\n") {
         if let Some(piece) = piece.strip_prefix('>') {
```

Comment:
what happens if there isn't a leading `>`? Should we silently ignore it or error out?

Reply:
If there isn't a leading `>` this code just silently ignores the `piece`. That's actually pretty useful in the manual. For example, you can make the tests skip a REPL statement, that I haven't bothered implementing; `:query` comes to mind.

Reply:
Okay, great. Just making sure it was intentional. That also might be worth a comment since I often encounter things like this that are silently ignoring errors.

maybe an empty `else` branch saying `// ignore unknown REPL statement e.g. :query`